pkg/querier/queryrange/roundtrip.go

```diff
@@ -46,12 +47,28 @@ type Config struct {
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.IntVar(&cfg.MaxRetries, "querier.max-retries-per-request", 5, "Maximum number of retries for a single request; beyond this, the downstream error is returned.")
 	f.BoolVar(&cfg.SplitQueriesByDay, "querier.split-queries-by-day", false, "Deprecated: Split queries by day and execute in parallel.")
-	f.DurationVar(&cfg.SplitQueriesByInterval, "querier.split-queries-by-interval", 0, "Split queries by an interval and execute in parallel, 0 disables it. You should use an a multiple of 24 hours (same as the storage bucketing scheme), to avoid queriers downloading and processing the same chunks.")
+	f.DurationVar(&cfg.SplitQueriesByInterval, "querier.split-queries-by-interval", 0, "Split queries by an interval and execute in parallel, 0 disables it. You should use an a multiple of 24 hours (same as the storage bucketing scheme), to avoid queriers downloading and processing the same chunks. This also determines how cache keys are chosen when result caching is enabled")
 	f.BoolVar(&cfg.AlignQueriesWithStep, "querier.align-querier-with-step", false, "Mutate incoming queries to align their start and end with their step.")
 	f.BoolVar(&cfg.CacheResults, "querier.cache-results", false, "Cache query results.")
 	cfg.ResultsCacheConfig.RegisterFlags(f)
 }
 
+func (cfg *Config) Validate() error {
+	if cfg.SplitQueriesByDay {
+		cfg.SplitQueriesByInterval = day
+	}
+
+	if cfg.CacheResults && cfg.SplitQueriesByInterval <= 0 {
+		return errors.New("querier.cache-results may only be enabled in conjunction with querier.split-queries-by-interval. Please set the latter")
+	}
+	return nil
+}
+
+// GenerateCacheKey impls CacheSplitter
+func (cfg Config) GenerateCacheKey(userID string, r Request) string {
```

Comment:
This is not a common pattern (having business logic functions to the Config). I personally did it some time ago and has been "rejected" by other maintainers, so I would get stick to it.

I would suggest to create `type cacheSplitter struct` taking in input the `cfg` and implementing `GenerateCacheKey()`. I would also remove the `generateKey()` function and move its logic into `cacheSplitter.GenerateCacheKey()` to simplify code reading.

The other place where we currently call `generateKey()` are its own tests, and will be easy to migrate that test to `cacheSplitter` test.